libyul/optimiser/UnusedFunctionParameterPruner.cpp

```diff
@@ -0,0 +1,324 @@
+
+	map<YulString, YulString> const& m_translations;
+	map<YulString, YulString> m_inverseTranslations;
+};
+
+vector<Statement> AddPrunedFunction::addFunction(FunctionDefinition& _old)
+{
+	auto generateName = [&](TypedName t)
+	{
+		return TypedName{
+			t.location,
+			m_nameDispenser.newName(t.name),
+			t.type
+		};
+	};
+
+	SourceLocation loc = _old.location;
+	YulString newName = m_inverseTranslations.at(_old.name);
+	TypedNameList functionArguments;
+	TypedNameList returnVariables;
+	TypedNameList renamedParameters = applyMap(_old.parameters, generateName);
+	TypedNameList reducedRenamedParameters;
+	TypedNameList renamedReturnVariables = applyMap(_old.returnVariables, generateName);
+	TypedNameList reducedRenamedReturnVariables;
+
+	if (m_argumentMask.count(newName))
+	{
+		vector<bool> const& mask = m_argumentMask.at(newName);
+		functionArguments = applyBooleanMask(_old.parameters, mask);
+		reducedRenamedParameters = applyBooleanMask(renamedParameters, mask);
+	}
+	else
+	{
+		functionArguments = _old.parameters;
+		reducedRenamedParameters = renamedParameters;
+	}
+
+	if (m_returnMask.count(newName))
+	{
+		vector<bool> const& mask = m_returnMask.at(newName);
+		returnVariables = applyBooleanMask(_old.returnVariables, mask);
+		reducedRenamedReturnVariables = applyBooleanMask(renamedReturnVariables, mask);
+	}
+	else
+	{
+		returnVariables = _old.returnVariables;
+		reducedRenamedReturnVariables = renamedReturnVariables;
+	}
+
+	FunctionDefinition newFunction{
+		loc,
+		move(newName),
+		move(functionArguments),
+		move(returnVariables),
+		{loc, {}} // body
+	};
+
+	swap(newFunction.body, _old.body);
+	swap(_old.parameters, renamedParameters);
+	swap(_old.returnVariables, renamedReturnVariables);
+
+	FunctionCall call{loc, Identifier{loc, newFunction.name}, {}};
+	for (auto const& p: reducedRenamedParameters)
+		call.arguments.emplace_back(Identifier{loc, p.name});
+
+	// Replace the body of `f_1` by an assignment which calls `f`, i.e.,
+	// `return_parameters = f(reduced_parameters)`
+	if (!newFunction.returnVariables.empty())
+	{
+		Assignment assignment;
+		assignment.location = loc;
+
+		// The LHS of the assignment.
+		for (auto const& r: reducedRenamedReturnVariables)
+			assignment.variableNames.emplace_back(Identifier{loc, r.name});
+
+		assignment.value = make_unique<Expression>(move(call));
+
+		_old.body.statements.emplace_back(move(assignment));
+	}
+	else
+		_old.body.statements.emplace_back(ExpressionStatement{loc, move(call)});
+
+	return make_vector<Statement>(move(newFunction), move(_old));
+}
+
+} // anonymous namespace
+
+void UnusedFunctionParameterPruner::run(OptimiserStepContext& _context, Block& _block)
+{
+	map<YulString, size_t> references =  ReferencesCounter::countReferences(_block);
+	FindFunctionsWithUnusedParameters find{_context.dialect, _block, references};
+
+	auto first = [](auto const& p) { return p.first; };
+	set<YulString> functionsWithUnusedArguments
```

Comment:
This is just the set of keys of the map, right? I think we should have a specific helper function for that.

Reply:
Yes. I wasn't able to find any helper for it. Maybe I'll just add it to `utils::`